src/kaleidoscope_internal/LEDModeManager.h

```diff
@@ -0,0 +1,428 @@
+
+// Generates a LEDModeFactory for each plugin. For all those
+// plugins that do not export dynamic LED modes, a nullptr-initialized
+// factory is created. This does not mean a performance penalty, as
+// those empty factories are then directly discarded, still at compile time.
+//
+#define _LED_MODE_MANAGEMENT__GENERATE_LED_MODE_FACTORY(PLUGIN)                \
+   kaleidoscope::internal::led_mode_management                          __NL__ \
+      ::GenerateLEDModeFactory<ledModePluginType(&::PLUGIN)>::apply(&::PLUGIN)
+
+// Retreive the pointer type of the exported dynamic LED mode of a plugin.
+//
+#define _LED_MODE_MANAGEMENT__PLUGIN_PTR_TYPE(PLUGIN)                          \
+   decltype(&::PLUGIN)
+
+#define _LED_MODE_MANAGEMENT__SETUP_STATIC_LED_MODE(PLUGIN)                    \
+   LEDModeManager::setupLEDMode(&::PLUGIN);
+
+#define _INIT_LED_MODE_MANAGER(...)                                            \
+   namespace kaleidoscope {                                             __NL__ \
+   namespace internal {                                                 __NL__ \
+   namespace led_mode_management {                                      __NL__ \
+                                                                        __NL__ \
+      /* Setup the array-like data structure that stores                __NL__ \
+       * LEDModeFactories */                                            __NL__ \
+      typedef kaleidoscope::internal::ArrayLikeStorage<                 __NL__ \
+               LEDModeFactory,                                          __NL__ \
+               /* Map the list of global plugin                         __NL__ \
+                * objects to a list of led mode                         __NL__ \
+                * factories */                                          __NL__ \
+               MAP_LIST(                                                __NL__ \
+                  /* Generate a series of boolean values. True for      __NL__ \
+                   * each plugin that is related to LED modes,          __NL__ \
+                   * false otherwise. */                                __NL__ \
+                  _LED_MODE_MANAGEMENT__PLUGIN_CONTROLS_LED_MODE,       __NL__ \
+                  __VA_ARGS__                                           __NL__ \
+               )                                                        __NL__ \
+      > ArrayType;                                                      __NL__ \
+                                                                        __NL__ \
+      /* Generate the actual instance of template class                 __NL__ \
+       * TypedPluginArray and initialize it with a list of              __NL__ \
+       * LEDModeFactories.                                              __NL__ \
+       */                                                               __NL__ \
+      const PROGMEM ArrayType led_mode_factories(                       __NL__ \
```

Comment:
>> That's why this PR does both, it moves LEDControl's LEDMode* array to PROGMEM and introduces dynamic LED modes.

> Would you mind pointing out where this array is in the code now? I can't find it.

The former `LEDMode` pointers, now `Plugin` pointers are stored in the `led_mode_factories` pseudo-array. For every LED mode related plugin instance the array stores a pointer to the instance and - if a dynamic LED mode - a pointer to a generator function (`generateLEDMode<...>`). This function generates the actual dynamic LED mode in the shared buffer. As it is templated, the factory function knows the exact type of the dynamic LED to create. For static LED modes, no factory function is required as the LED mode managing plugin is by itself a LED mode. See the implementation of `LEDModeManager::getLEDMode(...)` to see where this distinction is considered.

All this is an implementation of the [factory method](https://en.wikipedia.org/wiki/Factory_method_pattern) design pattern.